newgrp.1
.\" Copyright 1991, John F. Haugh II
.\" All rights reserved.
.\"
.\" Permission is granted to copy and create derivative works for any
.\" non-commercial purpose, provided this copyright notice is preserved
.\" in all copies of source code, or included in human readable form
.\" and conspicuously displayed on all copies of object code or
.\" distribution media.
.\"
.\"	@(#)newgrp.1	3.2	14:52:37	28 Apr 1992
.\"
.TH NEWGRP 1
.SH NAME
newgrp \- Change group ID
.br
sg \- Execute command as different group ID
.SH SYNOPSIS
.B newgrp
[ - ] [ \fIgroup\fR ]
.br
.B sg
[ - ] [ \fIgroup\fR [ \fB-c\fR \fIcommand\fR ] ]
.SH DESCRIPTION
.I newgrp
is used to change the current group ID during a login session.
If the optional \fB\-\fR flag is given, the user's environment
will be reinitialized as though the user had logged in, otherwise
the current environment, including current working directory,
remains unchanged.
.PP
.I newgrp
changes the current real group ID to the named group, or to
the default group listed in \fB/etc/passwd\fR if no group name
is given.
The user will be prompted for a password if they do not have a
password and the group does, or if the user is not listed as a
member and the group has a password.
The user will be denied access if the group password is empty
and the user is not listed as a member.
.PP
The
.I sg
command works similiar to \fInewgrp\fR but does not replace the
user's shell, so upon exit from a \fIsg\fR command, you are
returned to your previous group ID.
.I sg
also accepts a command.
The command will be executed with the Bourne shell and must be
enclosed in quotes.
.SH CAVEATS
.PP
This version of \fInewgrp\fR has many compilation options,
only some of which may be in use at any particular site.
.SH Files
/etc/passwd \- user account information
.br
/etc/group \- group information
.SH See Also
login(1),
id(1),
su(1)


NEWS
$Id$

shadow-4.1.5.1 -> shadow-4.2					UNRELEASED

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after
    sending a SIGTERM), and kill it only if it did not terminate by itself.
    No delay will be enforced if the child cooperates.
  * Default ENV_SUPATH is /sbin:/bin:/usr/sbin:/usr/bin
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.

newgrp.c
/*
 * Copyright 1990, 1991, 1992, 1993, John F. Haugh II
 * All rights reserved.
 *
 * Permission is granted to copy and create derivative works for any
 * non-commercial purpose, provided this copyright notice is preserved
 * in all copies of source code, or included in human readable form
 * and conspicuously displayed on all copies of object code or
 * distribution media.
 *
 * This software is provided on an AS-IS basis and the author makes
 * no warrantee of any kind.
 */

#include <sys/types.h>
#ifndef	BSD
#include <string.h>
#else
#include <strings.h>
#endif
#include <stdio.h>
#include <grp.h>
#include "pwd.h"
#include <termio.h>
#ifdef SYS3
#include <sys/ioctl.h>
#endif
#include "config.h"

#if !defined(BSD) && !defined(SUN) && !defined(SUN4)
#define	bzero(p,n) memset(p, 0, n)
#endif

#ifndef	lint
static	char	sccsid[] = "@(#)newgrp.c	3.11	07:34:23	08 Apr 1993";
#endif

#ifdef	NGROUPS
int	ngroups;
gid_t	groups[NGROUPS];
#endif

char	*getpass();
char	*getenv();
char	*pw_encrypt();
struct	passwd	*pwd;
struct	passwd	*getpwuid();
struct	passwd	*getpwnam();

#ifdef	SHADOWPWD
#include "shadow.h"
struct	spwd	*spwd;
struct	spwd	*getspnam();
#endif
#ifdef	SHADOWGRP
struct	sgrp	*sgrp;
struct	sgrp	*getsgnam();
#endif

#ifdef	USE_SYSLOG
#include <syslog.h>

/*VARARGS*/ int syslog();

#ifndef	LOG_WARN
#define	LOG_WARN LOG_WARNING
#endif	/* !LOG_WARN */
#endif	/* USE_SYSLOG */

struct	group	*grp;
struct	group	*getgrgid();
struct	group	*getgrnam();

char	*getlogin();
char	*crypt();
char	*getpass();
char	*getenv();
char	*pw_encrypt();
void	shell();

char	*name;
char	*group;
gid_t	gid;
int	cflag;

char	*Prog;
char	prog[BUFSIZ];
char	base[BUFSIZ];
char	passwd[BUFSIZ];
char	*cpasswd;
char	*salt;

#ifndef	MAXENV
#define	MAXENV	64
#endif

char	*newenvp[MAXENV];
int	newenvc = 0;
int	maxenv = MAXENV;

/*
 * usage - print command usage message
 */

usage ()
{
	if (strcmp (Prog, "sg") != 0)
		fprintf (stderr, "usage: newgrp [ - ] [ group ]\n");
	else
		fprintf (stderr, "usage: sg group [ command ]\n");
}

/*
 * newgrp - change the invokers current real and effective group id
 */

main (argc, argv, envp)
int	argc;
char	**argv;
char	**envp;
{
	int	initflag = 0;
	int	needspasswd = 0;
	int	i;
	char	*cp;
	char	*command;

	/*
	 * save my name for error messages and save my real gid incase
	 * of errors.  if there is an error i have to exec a new login
	 * shell for the user since her old shell won't have fork'd to
	 * create the process.  skip over the program name to the next
	 * command line argument.
	 */

	if (Prog = strrchr (argv[0], '/'))
		Prog++;
	else
		Prog = argv[0];

#ifdef	USE_SYSLOG
	openlog (Prog, LOG_PID|LOG_CONS|LOG_NOWAIT, LOG_AUTH);
#endif

	gid = getgid ();
	argc--; argv++;

	/*
	 * here i get to determine my current name.  i do this to validate
	 * my access to the requested group.  the validation works like
	 * this -
	 *	1) get the name associated with my current user id
	 *	2) get my login name, as told by getlogin().
	 *	3) if they match, my name is the login name
	 *	4) if they don't match, my name is the name in the
	 *	   password file.
	 *
	 * this isn't perfect, but it works more often then not.  i have
	 * to do this here so i can get the login name to find the
	 * login group.
	 */

	pwd = getpwuid (getuid ());

	if (! (name = getlogin ()) || strcmp (name, pwd->pw_name) != 0)
		name = pwd->pw_name;

	if (! (pwd = getpwnam (name))) {
		fprintf (stderr, "unknown user: %s\n", name);
#ifdef	USE_SYSLOG
		syslog (LOG_WARN, "unknown user `%s', uid `%d'\n",
			name, getuid ());
		closelog ();
#endif
		goto failure;
	}

	/*
	 * Parse the command line.  There are two accepted flags.  The
	 * first is "-", which for newgrp means to re-create the entire
	 * environment as though a login had been performed, and "-c",
	 * which for sg causes a command string to be executed.
	 *
	 * The next argument, if present, must be the new group name.
	 * Any remaining remaining arguments will be used to execute a
	 * command as the named group.  If the group name isn't present,
	 * I just use the login group ID of the current user.
	 *
	 * The valid syntax are
	 *	newgrp [ - ] [ groupid ]
	 *	sg [ - ]
	 *	sg [ - ] groupid [ command ]
	 */

	if (argc > 0 && argv[0][0] == '-' && argv[0][1] == '\0') {
		argc--; argv++;
		initflag = 1;
	}
	if (strcmp (Prog, "newgrp") != 0) {

		/* 
		 * Do the command line for everything that is
		 * not "newgrp".
		 */

		if (argc > 0 && argv[0][0] != '-') {
			group = argv[0];
			argc--; argv++;
		} else {
			usage ();
#ifdef	USE_SYSLOG
			closelog ();
#endif
			exit (1);
		}
		if (argc > 0) {
			command = argv[1];
			cflag++;
		}
	} else {

		/*
		 * Do the command line for "newgrp".  It's just
		 * making sure there aren't any flags and getting
		 * the new group name.
		 */

		if (argc > 0 && argv[0][0] == '-') {
			usage ();
			goto failure;
		} else if (argv[0] != (char *) 0) {
			group = argv[0];
		} else {

			/*
			 * get the group file entry for her login group id.
			 * the entry must exist, simply to be annoying.
			 */

			if (! (grp = getgrgid (pwd->pw_gid))) {
				fprintf (stderr, "unknown gid: %d\n",
					pwd->pw_gid);
#ifdef	USE_SYSLOG
				syslog (LOG_CRIT, "unknown gid: %d\n",
					pwd->pw_gid);
#endif
				goto failure;
			}
		}
	}
#ifdef	NGROUPS

	/*
	 * get the current users groupset.  the new group will be
	 * added to the concurrent groupset if there is room, otherwise
	 * you get a nasty message but at least your real and effective
	 * group id's are set.
	 */

	ngroups = getgroups (0, 0);
	if (ngroups > 0)
		getgroups (ngroups, groups);
#endif

	/*
	 * now we put her in the new group.  the password file entry for
	 * her current user id has been gotten.  if there was no optional
	 * group argument she will have her real and effective group id
	 * set to the value from her password file entry.  otherwise
	 * we validate her access to the specified group.
	 */

	if (group == (char *) 0) {
		if (! (grp = getgrgid (pwd->pw_gid))) {
			fprintf (stderr, "unknown gid: %d\n", pwd->pw_gid);
			goto failure;
		}
	} else if (! (grp = getgrnam (group))) {
		fprintf (stderr, "unknown group: %s\n", group);
		goto failure;
	}
#ifdef	SHADOWGRP
	sgrp = getsgnam (group);
#endif

	/*
	 * see if she is a member of this group.
	 */

	for (i = 0;grp->gr_mem[i];i++)
		if (strcmp (name, grp->gr_mem[i]) == 0)
			break;

	/*
	 * if she isn't a member, she needs to provide the
	 * group password.  if there is no group password, she
	 * will be denied access anyway.
	 */

	if (grp->gr_mem[i] == (char *) 0)
		needspasswd = 1;

#ifdef	SHADOWGRP
	if (sgrp) {

		/*
		 * Do the tests again with the shadow group entry.
		 */

		for (i = 0;sgrp->sg_mem[i];i++)
			if (strcmp (name, sgrp->sg_mem[i]) == 0)
				break;

		needspasswd = sgrp->sg_mem[i] == (char *) 0;
	}
#endif
#ifdef	SHADOWPWD

	/*
	 * if she does not have either a shadowed password,
	 * or a regular password, and the group has a password,
	 * she needs to give the group password.
	 */

	if (spwd = getspnam (name)) {
		if (spwd->sp_pwdp[0] == '\0' && grp->gr_passwd[0])
			needspasswd = 1;
#ifdef	SHADOWGRP
		if (spwd->sp_pwdp[0] == '\0' && sgrp != 0)
			needspasswd = sgrp->sg_passwd[0] != '\0';
#endif
	} else {
		if (pwd->pw_passwd[0] == '\0' && grp->gr_passwd[0])
			needspasswd = 1;
#ifdef	SHADOWGRP
		if (pwd->pw_passwd[0] == '\0' && sgrp != 0)
			needspasswd = sgrp->sg_passwd[0] != '\0';
#endif
	}
#else

	/*
	 * if she does not have a regular password she will have
	 * to give the group password, if one exists.
	 */

	if (pwd->pw_passwd[0] == '\0' && grp->gr_passwd[0])
		needspasswd = 1;
#ifdef	SHADOWGRP
	if (pwd->pw_passwd[0] == '\0' && sgrp != 0)
		needspasswd = sgrp->sg_passwd[0] != '\0';
#endif
#endif

	/*
	 * now i see about letting her into the group she requested.
	 * if she is the root user, i'll let her in without having to
	 * prompt for the password.  otherwise i ask for a password
	 * if she flunked one of the tests above.  note that she
	 * won't have to provide the password to her login group even
	 * if she isn't listed as a member.
	 */

	if (getuid () != 0 && needspasswd) {
		char	*encrypted;

		encrypted = grp->gr_passwd;
#ifdef	SHADOWGRP
		if (sgrp)
			encrypted = sgrp->sg_passwd;
#endif
		passwd[0] = '\0';

		if (encrypted[0]) {

		/*
		 * get the password from her, and set the salt for
		 * the decryption from the group file.
		 */

			if (! (cp = getpass ("Password:")))
				goto failure;

			strcpy (passwd, cp);
			bzero (cp, strlen (cp));
			salt = encrypted;
		} else {

		/*
		 * there is no password, print out "Sorry" and give up
		 */

			fputs ("Sorry\n", stderr);
			goto failure;
		}

		/*
		 * encrypt the key she gave us using the salt from
		 * the password in the group file.  the result of
		 * this encryption must match the previously
		 * encrypted value in the file.
		 */

		cpasswd = pw_encrypt (passwd, salt);
		bzero (passwd, sizeof passwd);

		if (strcmp (cpasswd, encrypted) != 0) {
			fputs ("Sorry\n", stderr);
#ifdef	USE_SYSLOG
		syslog (LOG_INFO, "Invalid password for `%s' from `%s'\n",
			group, name);
#endif
			goto failure;
		}
	}

	/*
	 * all successful validations pass through this point.  the
	 * group id will be set, and the group added to the concurrent
	 * groupset.
	 */

#ifdef	USE_SYSLOG
	if (getdef_bool ("SYSLOG_SU_ENAB"))
		syslog (LOG_INFO, "user `%s' switched to group `%s'\n", name, group);
#endif
	gid = grp->gr_gid;
#ifdef	NGROUPS

	/*
	 * i am going to try to add her new group id to her concurrent
	 * group set.  if the group id is already present i'll just
	 * skip this part.  if the group doesn't fit, i'll complain
	 * loudly and skip this part ...
	 */

	for (i = 0;i < ngroups;i++) {
		if (gid == groups[i])
			break;
	}
	if (i == ngroups) {
		if (ngroups == NGROUPS) {
			fprintf (stderr, "too many groups\n");
		} else {
			groups[ngroups++] = gid;
			if (setgroups (ngroups, groups)) {
				fprintf (stderr, "%s: ", Prog);
				perror ("unable to set groups");
			}
		}
	}
#endif

okay:

	/*
	 * i set her group id either to the value she requested, or
	 * to the original value if the newgrp failed.
	 */

	if (setgid (gid))
		perror ("setgid");

	if (setuid (getuid ()))
		perror ("setuid");

	/*
	 * see if the "-c" flag was used.  if it was, i just create a
	 * shell command for her using the argument that followed the
	 * "-c" flag.
	 */

	if (cflag) {
		execl ("/bin/sh", "sh", "-c", command, (char *) 0);
		perror ("/bin/sh");
#ifdef	USE_SYSLOG
		closelog ();
#endif
		exit (255);
	}

	/*
	 * i have to get the pathname of her login shell.  as a favor,
	 * i'll try her environment for a $SHELL value first, and
	 * then try the password file entry.  obviously this shouldn't
	 * be in the restricted command directory since it could be
	 * used to leave the restricted environment.
	 */

	if (! initflag && (cp = getenv ("SHELL")))
		strncpy (prog, cp, sizeof prog);
	else if (pwd->pw_shell && pwd->pw_shell[0])
		strncpy (prog, pwd->pw_shell, sizeof prog);
	else
		strcpy (prog, "/bin/sh");

	/*
	 * now i try to find the basename of the login shell.  this
	 * will become argv[0] of the spawned command.
	 */

	if (cp = strrchr (prog, '/'))
		cp++;
	else
		cp = prog;

	/*
	 * to have the shell perform login processing i will set the
	 * first character in the first argument to a "-".
	 */

	if (initflag)
		strcat (strcpy (base, "-"), cp);
	else
		strcpy (base, cp);

#ifdef	SHADOWPWD
	endspent ();
#endif
#ifdef	SHADOWGRP
	endsgent ();
#endif
	endpwent ();
	endgrent ();

	/*
	 * switch back to her home directory if i am doing login
	 * initialization.
	 */

	if (initflag) {
		chdir (pwd->pw_dir);
		while (*envp) {
			if (strncmp (*envp, "PATH=", 5) == 0 ||
					strncmp (*envp, "HOME=", 5) == 0 ||
					strncmp (*envp, "SHELL=", 6) == 0 ||
					strncmp (*envp, "TERM=", 5) == 0)
				addenv (*envp);

			envp++;
		}
	} else {
		while (*envp)
			addenv (*envp++);
	}

	/*
	 * exec the login shell and go away.  we are trying to get
	 * back to the previous environment which should be the
	 * user's login shell.
	 */

	shell (prog, base);
	/*NOTREACHED*/

failure:
	/*
	 * this is where all failures land.  the group id will not
	 * have been set, so the setgid() below will set me to the
	 * original group id i had when i was invoked.
	 */

	/*
	 * only newgrp needs to re-exec the user's shell.  that is
	 * because the shell doesn't recognize "sg", so it doesn't
	 * "exec" this command.
	 */

	if (strcmp (Prog, "newgrp") != 0) {
#ifdef	USE_SYSLOG
		closelog ();
#endif
		exit (1);
	}
	
	/*
	 * The GID is still set to the old value, so now I can
	 * give the user back her shell.
	 */

	goto okay;
}
